wire-runtime/src/main/java/com/squareup/wire/RuntimeMessageAdapter.java

```diff
@@ -53,20 +51,69 @@
     this.builderType = getBuilderType(messageType);
     this.builderCopyConstructor = getBuilderCopyConstructor(builderType, messageType);
 
-    Map<Integer, RuntimeFieldBinding> map = new LinkedHashMap<Integer, RuntimeFieldBinding>();
+    Map<Integer, TagBinding<M, Builder<M>>> tagBindings
+        = new LinkedHashMap<Integer, TagBinding<M, Builder<M>>>();
+
+    // Create tag bindings for fields annotated with '@ProtoField'
     for (Field messageField : messageType.getDeclaredFields()) {
-      // Process fields annotated with '@ProtoField'
       ProtoField protoField = messageField.getAnnotation(ProtoField.class);
       if (protoField != null) {
-        int tag = protoField.tag();
-        map.put(tag, RuntimeFieldBinding.create(wire, protoField, messageField, builderType));
+        TypeAdapter<?> singleAdapter = singleTypeAdapter(wire, messageField, protoField);
+        Class<?> singleType = singleAdapter.javaType;
+        FieldTagBinding<M> tagBinding = new FieldTagBinding<M>(protoField,
+            protoField.type(), singleAdapter, singleType, messageField, builderType);
+        tagBindings.put(tagBinding.tag, tagBinding);
       }
     }
-    fieldBindingMap = TagMap.of(map);
+
+    // Create tag bindings for registered extensions.
+    for (Map.Entry<Integer, Extension<?, ?>> entry
+        : wire.registry.getExtensions(messageType).entrySet()) {
+      Extension<?, ?> extension = entry.getValue();
+      TypeAdapter<?> singleAdapter = TypeAdapter.get(wire, extension.getDatatype(),
+          extension.getMessageType(), extension.getEnumType());
+      tagBindings.put(entry.getKey(), new ExtensionTagBinding<M>(extension, singleAdapter));
+    }
+
+    this.tagBindings = Collections.unmodifiableMap(tagBindings);
```

Comment:
This is the magic: a single map that contains both message tags and extension tags.